r_d15/r_d15.cpp

```diff
@@ -0,0 +1,120 @@
+#include <iostream>
+#include <string>
+
+/*
+�������� ��������� Student � ������ name �� marks[4];
+
+�������� �������, �� ������ �������� �� ������� �������� �� ������� ������� ������ �� ��������� ��������;
+�������� ����� of Student;
+�������, �� ����� �������� � ����� �� ������� �������� ������;
+�������, �� ������� ���������(pointer) �� ������������� �������� � �����;
+�������, ��� �������� ������� ��������, ������� ������ ���� ����� ��� ������� �������� N;
+�������, ��� ������� ����� � N% ������������ ��������:
+void getBestStudents(Student* inStudents, unsigned inSize, Student* outStudents, unsigned& outSize, unsigned percent);
+*/
+
+constexpr short MARKS_COUNT = 4;
+constexpr short STUDENTS_COUNT = 4;
+
+struct Student
+{
+    std::string name;
+    unsigned short marks[MARKS_COUNT];
+};
+
+double getAverage(Student student) {
+    double res = 0;
+    for (int i = 0; i < MARKS_COUNT; i++) {
+        res += student.marks[i];
+    }
+    return res / MARKS_COUNT;
+}
+
+void sortStudentsByAvarage(Student* students, unsigned short studentsCount) {
+    Student tmp;
+    for (int i = 0; i < studentsCount - 1; i++) {
+        for (int j = 0; j < studentsCount - i - 1; j++) {
+            if (getAverage(students[j]) > getAverage(students[j + 1])) {
+                tmp = students[j];
+                students[j] = students[j + 1];
+                students[j + 1] = tmp;
+            }
+        }
+    }
+}
+
+Student* getTheBestStudent(Student* students, unsigned short studentsCount) {
+    int theBestIndex = 0;
+    for (int i = 0; i < studentsCount - 1; i++) {
+        if (getAverage(students[i+1]) > getAverage(students[i])) {
+            theBestIndex = i;
+        }
+    }
+
+    return &students[theBestIndex];
+}
+
+unsigned countStudentsBetterThanN(Student* students, unsigned short studentsCount, double N) {
+    unsigned counter = 0;
+    for (int i = 0; i < studentsCount; i++) {
+        if (getAverage(students[i]) > N) {
+            counter++;
+        }
+    }
+
+    return counter;
+}
+
+void getBestStudents(Student* inStudents, unsigned inSize, Student*& outStudents, unsigned& outSize, unsigned percent) {
+    outSize = static_cast<unsigned>((inSize * percent) / 100.0);
+
+    outStudents = new Student[outSize];
+    sortStudentsByAvarage(inStudents, inSize);
```

Comment:
sorting is not the best idea here. Can be implemented faster than N*logN of sorting with simple search